### PR TITLE
[Process] Add support for Fiber

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        |

---

Example:

(Note: The loop is very naive, but it's for demo purpose)

```php
<?php

use Symfony\Component\Process\Process;

require __DIR__ . '/vendor/autoload.php';

function logg(string $message, string ...$params)
{
    dump(sprintf($message, ...$params));
}

class Loop
{
    public function __construct(
        private array $fibers
    ) {
    }

    public function run()
    {
        foreach ($this->fibers as $k => $fiber) {
            logg('Fiber %d started', $k);
            $fiber->start();
        }

        while ($this->fibers) {
            foreach ($this->fibers as $k => $fiber) {
                if ($fiber->isTerminated()) {
                    logg('Fiber %d finished', $k);
                    unset($this->fibers[$k]);
                    continue;
                }

                $fiber->resume();
            }

            usleep(100_000);
        }
    }
}

// // Sync
// $process = new Process(['sleep', 1]);
// $process->mustRun();

// Async
$fiber1 = new Fiber(function () {
    $process = new Process(['sleep', 1]);
    $process->mustRun();
});
$fiber2 = new Fiber(function () {
    $process = new Process(['sleep', 2]);
    $process->mustRun();
});
$fiber3 = new Fiber(function () {
    $process = new Process(['sleep', 3]);
    $process->mustRun();
});

// dump($fiber1);die();


$loop = new Loop([$fiber3, $fiber2, $fiber1]);
$loop->run();
```

https://user-images.githubusercontent.com/408368/138566377-735277f3-c27f-4b69-a15d-04c0a8c9b29b.mp4

### CODE REVIEW

File: src/Symfony/Component/Process/CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,10 @@
 CHANGELOG
 =========
+5.4.0
+
+-----
```
Comment (derrabus) [inline src/Symfony/Component/Process/CHANGELOG.md]:
```suggestion

5.4
---
```

Comment (lyrixx) [reply src/Symfony/Component/Process/CHANGELOG.md]: thanks, fixed.  for the `5.4.0` => I followed the previous nomenclature

**Comment by nicolas-grekas** [comment]
Nice, can you add a test case maybe ? That'd help understanding how to use it or just that it's possible.

**Comment by lyrixx** [comment]
I added a test case 👍🏼

**Comment by b-viguier** [comment]
Glad to see that you consider using Fibers! 🤗 
Just some thoughts about this:
* What if the fiber is never resumed?
* What if the fiber is resumed with `throw`?
* What if the fiber is resumed without calling `usleep` in the meantime? (To be honest, I'm not sure `Process` actually needs this…)

My point is that it could be interesting to define a _contract_ about _"how to use safely a SF Process in a Fiber"_.
A part of this contract could also be transmitted via the `suspend` function, a kind of _promise_ to be resumed in `n` microseconds, or when a goal is reached… 🤷 

I have no particular idea about this, I'm not sure if Symfony needs a complete async tools set, but I'm just wondering if this `suspend` can be an open door to wrong usages.

Anyway, this PR has my full attention 😉👍

**Comment by lyrixx** [comment]
> Glad to see that you consider using Fibers!

Thanks a lot for your feedback

> * What if the fiber is never resumed?

The process will last, but it's exactly the same as when you write the follow code
```
$p = new Process['sleep', '10'];
$p->start();
// Nothing here
```

So IMHO, nothing has to be done here

> * What if the fiber is resumed with `throw`?

What would be the use case to resume it with `throw`?

> * What if the fiber is resumed without calling `usleep` in the meantime? (To be honest, I'm not sure `Process` actually needs this…)

Indeed, I was not sure about the use case, but actually, it eases the async handling of process.


> My point is that it could be interesting to define a _contract_ about _"how to use safely a SF Process in a Fiber"_.

I totally agree with that 👍🏼, and more generally in Symfony

>  I'm just wondering if this `suspend` can be an open door to wrong usages.

Yes, it can. I said that, because we keeps reading issue where people do really strange things :) I guess it's how OSS works :)

**Comment by b-viguier** [comment]
> The process will last, but it's exactly the same as when you write the follow code [...]
> So IMHO, nothing has to be done here

👍 


> What would be the use case to resume it with `throw`?

Since it's possible, we can expect that someone will try 😅.
But it could a way to stop a too long process (timeout), or to cancel it for other external reasons (too much opened fibers, too much memory... 🤷‍♂️)


> Yes, it can. I said that, because we keeps reading issue where people do really strange things :) I guess it's how OSS works :) 

😅😉

File: src/Symfony/Component/Process/Process.php [hunk 2]
```diff
@@ -427,12 +427,22 @@ public function wait(callable $callback = null)
         do {
             $this->checkTimeout();
             $running = '\\' === \DIRECTORY_SEPARATOR ? $this->isRunning() : $this->processPipes->areOpen();
-            $this->readPipes($running, '\\' !== \DIRECTORY_SEPARATOR || !$running);
+
+            if (\PHP_VERSION_ID >= 80100 && $fiber = \Fiber::getCurrent()) {
```
Comment (joelwurtz) [inline src/Symfony/Component/Process/Process.php]: not sure about this check, fiber is also available in 8.0 with an extension (https://github.com/amphp/ext-fiber)

File: src/Symfony/Component/Process/CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+5.4
+---
+
+ * added support for `Fiber`
```
Comment (OskarStark) [inline src/Symfony/Component/Process/CHANGELOG.md]:
```suggestion
 * Add support for `Fiber`
```

File: src/Symfony/Component/Process/Tests/ProcessTest.php [hunk 4]
```diff
@@ -550,6 +550,30 @@ public function testMustRun()
         $this->assertEquals('foo'.\PHP_EOL, $process->getOutput());
     }
 
+    /**
+     * @requires PHP 8.1
```
Comment (OskarStark) [inline src/Symfony/Component/Process/Tests/ProcessTest.php]:
```suggestion
     * @requires ext-fiber
```
?

Comment (nicolas-grekas) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: won't work, extension_loaded('fiber') returns false on 8.1 if I tested correctly

Comment (OskarStark) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: Weird, but isn't this available in 8.0 as an extension too? 
I could be wrong though...

Comment (lyrixx) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: Indeed, this is weird:
<details>

```
php > ^D>/home/gregoire php -v
PHP 8.1.0RC4 (cli) (built: Oct 14 2021 16:10:05) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.0RC4, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.0RC4, Copyright (c), by Zend Technologies
>/home/gregoire php -m
[PHP Modules]
bcmath
calendar
Core
ctype
curl
date
dom
exif
FFI
fileinfo
filter
ftp
gd
gettext
hash
iconv
intl
json
libxml
mbstring
mysqli
mysqlnd
openssl
pcntl
pcre
PDO
pdo_mysql
pdo_pgsql
pdo_sqlite
pgsql
Phar
posix
readline
Reflection
session
shmop
SimpleXML
soap
sockets
sodium
SPL
sqlite3
standard
sysvmsg
sysvsem
sysvshm
tokenizer
xml
xmlreader
xmlwriter
xsl
Zend OPcache
zip
zlib

[Zend Modules]
Zend OPcache

>/home/gregoire php -a
Interactive shell

php > var_dump(Fiber::getCurrent());
NULL
php > 
```

</details>

TL;DR: on PHP 8.1 I do have fiber, but the extension is not listed

Comment (nicolas-grekas) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: worth a bug report I suppose

Comment (stof) [reply src/Symfony/Component/Process/Tests/ProcessTest.php]: Looking at the php source code, the Fiber class is declared as part of the [Zend engine](https://github.com/php/php-src/blob/578b785ecff5a8ba4a3784679e6c4d7e5f11bc20/Zend/zend_fibers.c), not as part of a core extension.

File: src/Symfony/Component/Process/Process.php [hunk 5]
```diff
@@ -427,12 +427,22 @@ public function wait(callable $callback = null)
...
+            } else {
+                $this->readPipes($running, '\\' !== \DIRECTORY_SEPARATOR || !$running);
+            }
         } while ($running);
 
         while ($this->isRunning()) {
             $this->checkTimeout();
-            usleep(1000);
+            if (\PHP_VERSION_ID >= 80100 && $fiber = \Fiber::getCurrent()) {
+                $fiber->suspend();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Process.php:449]: This relies on the idea that the underlying scheduler is going to sleep when it has nothing to do. But nothing requires that.

We could sleep for 1000 minus the time spent suspended to cover that, don't you think?

File: src/Symfony/Component/Process/Process.php (hunk 2 above)
Comment (lyrixx) [reply src/Symfony/Component/Process/Process.php]: I replaced it with a `class_exist()`

File: src/Symfony/Component/Process/CHANGELOG.md (hunk 3 above)
Comment (lyrixx) [reply src/Symfony/Component/Process/CHANGELOG.md]: Thanks

File: src/Symfony/Component/Process/Process.php (hunk 5 above)
Comment (lyrixx) [reply src/Symfony/Component/Process/Process.php:449]: I added a safe guard 👍🏼

**Comment by fabpot** [comment]
What's the status here? @lyrixx?

**Comment by lyrixx** [comment]
I think there is consensus yet on how to add fibrer support in Symfony. Let's close it

But if someone want it, we could also merge it.

**Comment by soyuka** [comment]
mhh wait why is this not merged?

**Comment by lyrixx** [comment]
Wep, too bad. People are not ready for that! 

And we need that in [castor](https://github.com/jolicode/castor/blob/main/src/functions.php#L205-L211)

cc @joelwurtz

**Comment by fabpot** [comment]
@lyrixx Not sure how you came to that conclusion as you were the one closing this PR.

**Comment by lyrixx** [comment]
Let's re-open the PR 👍🏼 (rebased, and updated)

File: src/Symfony/Component/Process/Process.php [hunk 6]
```diff
@@ -429,12 +430,31 @@ public function wait(?callable $callback = null): int
         do {
             $this->checkTimeout();
             $running = $this->isRunning() && ('\\' === \DIRECTORY_SEPARATOR || $this->processPipes->areOpen());
-            $this->readPipes($running, '\\' !== \DIRECTORY_SEPARATOR || !$running);
+            if (class_exists(\Fiber::class, false) && $fiber = \Fiber::getCurrent()) {
```
Comment (derrabus) [inline src/Symfony/Component/Process/Process.php]: Doesn't the `Fiber` class always exist since PHP 8.1?

File: src/Symfony/Component/Process/Process.php [hunk 7]
```diff
@@ -11,6 +11,7 @@
 
 namespace Symfony\Component\Process;
 
+use Symfony\Component\DependencyInjection\Tests\Compiler\E;
```
Comment (derrabus) [inline src/Symfony/Component/Process/Process.php]: ?

Comment (lyrixx) [reply src/Symfony/Component/Process/Process.php]: oups

**Review by dunglas — APPROVED** [review]

File: src/Symfony/Component/Process/Process.php [hunk 8]
```diff
@@ -429,12 +429,31 @@ public function wait(?callable $callback = null): int
         do {
             $this->checkTimeout();
             $running = $this->isRunning() && ('\\' === \DIRECTORY_SEPARATOR || $this->processPipes->areOpen());
-            $this->readPipes($running, '\\' !== \DIRECTORY_SEPARATOR || !$running);
+            if ($fiber = \Fiber::getCurrent()) {
+                $this->readPipes(false, '\\' !== \DIRECTORY_SEPARATOR || !$running);
+                $startedAt = microtime(true);
+                $fiber->suspend();
+                $sleepFor = (int) (1000 - (microtime(true) - $startedAt) * 1000000);
+                if (0 < $sleepFor) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Process/Process.php:437]: (same below)
```suggestion
                $suspendedAt = microtime(true);
                $fiber->suspend();
                if (0 < $sleepFor = (int) (1000 - (microtime(true) - $suspendedAt) * 1000000)) {
```

But I would also make readPipe return true/false whether the process did yield some activity or not. This way, we won't force a sleep while there are things to do.

**Comment by joelwurtz** [comment]
After playing a bit with the process and fiber, i think there is no need to add fiber support in this class, the only things the process class need would be getter for the `stdout` and `stderr` stream.

Here is an example on how it may work (i use closure binding to get the streams) : 

```php
require 'vendor/autoload.php';

$pid = null;
$running = true;

// listen for sigchld
pcntl_signal(SIGCHLD, function ($signo, $data) use (&$pid, &$running) {
    if ($pid === $data['pid']) {
        $running = false;
    }
});

pcntl_async_signals(true);

$fiber = new Fiber(function () {
    $process = \Symfony\Component\Process\Process::fromShellCommandline('echo -n "test"; echo -n "bar" 1>&2 ; sleep 2');

    // bind a function to get the output stream
    $getStdout = Closure::bind(function () {
        return $this->stdout;
    }, $process, $process);
    $getStderr = Closure::bind(function () {
        return $this->stderr;
    }, $process, $process);

    $process->start();

    $stdout = $getStdout();
    $stderr = $getStderr();

    stream_set_blocking($stdout, 0);
    stream_set_blocking($stderr, 0);

    while ($process->isRunning()) {
        $action = \Fiber::suspend([$stdout, $stderr]);

        if ($action === 'stdout') {
            $content = $process->getIncrementalOutput();

            if ($content !== '') {
                var_dump('stdout : ' . $content);
            }
        }

        if ($action === 'stderr') {
            $content = $process->getIncrementalErrorOutput();

            if ($content !== '') {
                var_dump('stderr : ' . $content);
            }
        }
    }

    return $process->getExitCode();
});

[$stdout, $stderr] = $fiber->start();
$streams = [$stdout, $stderr];

while (!$fiber->isTerminated()) {
    $write = null;
    $except = null;

    $read = $streams;
    $modified = stream_select($read, $write, $except, 0, 100_000);

    if ($modified === false) {
        break;
    }

    if ($modified > 0) {
        foreach ($read as $stream) {
            if ($fiber->isTerminated()) {
                break;
            }

            if ($stream === $stdout) {
                $fiber->resume('stdout');
            }

            if ($stream === $stderr) {
                $fiber->resume('stderr');
            }
        }
    }
}

var_dump('exit code : ' . $fiber->getReturn());
```

There may be some adjustement to do in the process class, but checking for fibers may not be needed IMHO

Also this totally remove sleep inside the process, (it may sleep but it depends only on the event loop implementation)

Exposing those streams make it be compatible with [Revolt](https://revolt.run/) also, and i believe it would be better to provider a helper around the `Process` class that make it compatible with revolt ?

**Comment by joelwurtz** [comment]
Didn't see this was in the `wait` function, it can make sense for it for existing library / app that use this method.

Part of this example can be reuse, like instead of doing a `sleep` it could do a stream_select with a timeout so it wakes up as soon as there is something to read / write or use the `SIGCHLD` signal to wake up also when the process has stopped (only with pcntl extension) ?

File: src/Symfony/Component/Process/Process.php [hunk 9]
```diff
@@ -429,12 +429,31 @@ public function wait(?callable $callback = null): int
         do {
             $this->checkTimeout();
             $running = $this->isRunning() && ('\\' === \DIRECTORY_SEPARATOR || $this->processPipes->areOpen());
-            $this->readPipes($running, '\\' !== \DIRECTORY_SEPARATOR || !$running);
+            if ($fiber = \Fiber::getCurrent()) {
```
Comment (joelwurtz) [inline src/Symfony/Component/Process/Process.php:432]: I believe there should be an additional flag to enable this support otherwise it would be a bc break for people already using this method in a `Fiber`

File: src/Symfony/Component/Process/Process.php [hunk 10]
```diff
@@ -429,12 +429,31 @@ public function wait(?callable $callback = null): int
...
+                $this->readPipes(false, '\\' !== \DIRECTORY_SEPARATOR || !$running);
+                $startedAt = microtime(true);
+                $fiber->suspend();
+                $sleepFor = (int) (1000 - (microtime(true) - $startedAt) * 1000000);
+                if (0 < $sleepFor) {
+                    usleep($sleepFor);
```
Comment (joelwurtz) [inline src/Symfony/Component/Process/Process.php:438]: maybe it could use stream_select on the pipes instead of usleep ? so it would directly read / write if there is something to do instead of sleeping ?

Comment (nicolas-grekas) [reply src/Symfony/Component/Process/Process.php:438]: we do use stream_select already so I guess we added this to address a situation where stream_select doesn't apply. git blame could help remember :)

**Comment by fabpot** [comment]
Is it still relevant?
